File: Algorithm_Solution/baekj/bk_1647.py
# # 도시분할계획
# # prim 4000ms
# import sys
# from heapq import heappop, heappush
#
# # [A] 입력 함수 초기화
# def input():
#     return sys.stdin.readline().rstrip('\n')
#
# # [B] prim
# def prim(N, adjL):
#     answer = 0
#     q = [(0, 1)]
#     maxV = cnt = 0
#     connected = [False] * (N + 1)
#     while q:
#         cost, v = heappop(q)
#         if connected[v]: continue
#         connected[v] = True
#         answer += cost
#         maxV = max(maxV, cost)  # 가장 비싼 간선 저장
#         cnt += 1
#         if cnt >= N - 2: break
#         for c, w in adjL[v]:
#             if connected[w]: continue
#             heappush(q, (c, w))
#
#     # 선택된 MST 간선중 가장 비용이 비싼 간선의 비용 빼기
#     return answer - maxV
#
# # [main]
# if __name__ == '__main__':
#     N, M = map(int, input().split())
#     adjL = [[] for _ in range(N + 1)]
#     for _ in range(M):
#         a, b, c = map(int, input().split())
#         adjL[a].append((c, b))
#         adjL[b].append((c, a))
#     print(prim(N, adjL))

# 도시분할계획
# kruskal 2500ms
import sys

# [A] 입력 함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [B] 부모 노드를 찾는 함수
def find(x, par):
    while x != par[x]:
        x = par[par[x]]
    return x

# [C] 부모노드를 병합하는 함수
def union(x, y, par):
    x = find(x, par)
    y = find(y, par)
    par[min(x, y)] = max(x, y)

# [D] MST 크루스칼
def kruskal(N, edges):
    edges.sort(key=lambda x: x[2])
    par = list(range(N + 1))
    cnt = answer = 0
    for a, b, c in edges:
        if cnt >= N - 2: break
        if find(a, par) != find(b, par):
            union(a, b, par)
            cnt += 1
            answer += c
    return answer

# [main]
if __name__ == '__main__':
    N, M = map(int, input().split())
    edges = [tuple(map(int, input().split())) for _ in range(M)]
    print(kruskal(N, edges))

File: BOJ_JS/그래프/MST/bk_1647.js
/**
 * 제목 : 도시분할 계획 - 백준(골드 4)
 * 분류 : MST, 그래프 이론
 */

/**
 * 접근 방식
 * 1. 노드 N개, 양방향 간선 M개, 가중치 C
 * 2. 간선 중심의 크루스칼 + union/find
 */

const solution = (input) => {
  let line = 0;
  const [N, M] = input[line++].split(" ").map(v => Number(v));
  let edges = [];
  while (line < M + 1) {
    edges.push(input[line++].split(" ").map(v => Number(v)));
  }

  edges.sort((a, b) => a[2] - b[2]);
  let par = Array.from({ length: N + 1 }, (_, i) => i);
  let [cnt, answer] = [0, 0];
  for (let [a, b, c] of edges) {
    if (cnt >= N - 2) break;
    if (find(a) != find(b)) {
      union(a, b);
      cnt += 1;
      answer += c;
    }
  }

  return answer;

  function find(x) {
    while (x != par[x]) {
      x = par[par[x]];
    }
    return x;
  };

  function union(x, y) {
    x = find(x, par);
    y = find(y, par);
    par[Math.min(x, y)] = Math.max(x, y);
  }
};

const input = require('fs').readFileSync('./input.txt').toString().split('\n');
console.log(solution(input));